Strip out the non-indexed fields before passing the JSON instance to ES for processing

When I wrote [tests](https://github.com/MurmurationsNetwork/MurmurationsServices/pull/557/files#diff-f198d179caabce55f90e9323c320240d691f7b151b7995ba8ad510cffd8b50a4R212) for the schemavalidator, I noticed that when I provide unspecified data, it can pass the validation. I think we should not allow users to add random data, and we can prevent this by adding the [additionalProperties](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties): false attribute to the schema definition.

```
--- FAIL: TestValidate (0.01s)
    --- FAIL: TestValidate/Profile_with_Unspecified_Data (0.00s)
        /Users/max/Desktop/MurmurationsServices/pkg/schemavalidator/schemavalidator_test.go:289: 
            	Error Trace:	/Users/max/Desktop/MurmurationsServices/pkg/schemavalidator/schemavalidator_test.go:289
            	Error:      	Not equal: 
            	            	expected: false
            	            	actual  : true
            	Test:       	TestValidate/Profile_with_Unspecified_Data
FAIL
FAIL	github.com/MurmurationsNetwork/MurmurationsServices/pkg/schemavalidator	0.378s
```

--- a/services/index/internal/model/export_test.go
+++ b/services/index/internal/model/export_test.go
@@ -19,12 +19,6 @@ func (s *TestProfile) ConvertGeolocation() error {
 	return s.convertGeolocation()
 }
 
-// RepackageGeolocation is a wrapper around the unexported repackageGeolocation
-// method in the Profile type.
-func (s *TestProfile) RepackageGeolocation() {
-	s.repackageGeolocation()
-}
-
 // NormalizeCountryCode is a wrapper around the unexported normalizeCountryCode
 // method in the Profile type.
 func (s *TestProfile) NormalizeCountryCode() error {

--- a/services/index/internal/model/profile.go
+++ b/services/index/internal/model/profile.go
@@ -15,6 +15,24 @@ import (
 	"github.com/MurmurationsNetwork/MurmurationsServices/services/index/config"
 )
 
+// AllowedFields specifies indexable keys.
+//
+// Elasticsearch stores the original JSON document, and to prevent the index
+// from containing garbage data, we manually filter out unwanted fields.
+var AllowedFields = map[string]bool{
+	"country":        true,
+	"geolocation":    true,
+	"last_updated":   true,
+	"linked_schemas": true,
+	"locality":       true,
+	"name":           true,
+	"primary_url":    true,
+	"profile_url":    true,
+	"region":         true,
+	"status":         true,
+	"tags":           true,
+}
+
 // Profile represents the profile data for a node.
 type Profile struct {
 	// Original profile string.
@@ -33,9 +51,17 @@ func NewProfile(profileStr string) *Profile {
 	}
 }
 
-// GetJSON returns the JSON representation of the profile.
+// GetJSON returns the filtered JSON representation of the profile.
 func (p *Profile) GetJSON() map[string]interface{} {
-	return p.json
+	filteredJSON := make(map[string]interface{})
+
+	for key, value := range p.json {
+		if _, ok := AllowedFields[key]; ok {
+			filteredJSON[key] = value
+		}
+	}
+
+	return filteredJSON
 }
 
 // Update processes and updates the profile data.

--- a/services/index/internal/model/profile.go
+++ b/services/index/internal/model/profile.go
@@ -2,6 +2,7 @@ package model
 
 import (
 	"errors"
+	"fmt"
 	"strconv"
 	"strings"
 
@@ -75,7 +76,6 @@ func (p *Profile) Update(
 	if err := p.convertGeolocation(); err != nil {
 		return err
 	}
-	p.repackageGeolocation()
 
 	if err := p.normalizeCountryCode(); err != nil {
 		return err
@@ -90,43 +90,45 @@ func (p *Profile) Update(
 	return nil
 }
 
-// convertGeolocation parses a string-formatted geolocation into latitude and
-// longitude values. The geolocation should be in the format "latitude,longitude".
+// convertGeolocation parses a geolocation string into a specific format.
 func (p *Profile) convertGeolocation() error {
-	if geo, ok := p.json["geolocation"].(string); ok {
-		g := strings.Split(geo, ",")
-		var err error
-		p.json["latitude"], err = strconv.ParseFloat(g[0], 64)
-		if err != nil {
-			return err
+	var lat, lon float64
+	var err error
+
+	geoLocation := map[string]interface{}{"lat": 0, "lon": 0}
+
+	if geoStr, ok := p.json["geolocation"].(string); ok {
+		g := strings.Split(geoStr, ",")
+		if len(g) != 2 {
+			return fmt.Errorf("invalid geolocation format")
 		}
-		p.json["longitude"], err = strconv.ParseFloat(g[1], 64)
+
+		lat, err = strconv.ParseFloat(g[0], 64)
 		if err != nil {
-			return err
+			return fmt.Errorf("invalid latitude: %v", err)
+		}
+
+		lon, err = strconv.ParseFloat(g[1], 64)
+		if err != nil {
+			return fmt.Errorf("invalid longitude: %v", err)
+		}
+
+		geoLocation["lat"] = lat
+		geoLocation["lon"] = lon
+	} else if existingGeo, ok := p.json["geolocation"].(map[string]interface{}); ok {
+		geoLocation = existingGeo
+	} else {
+		if existingLat, ok := p.json["latitude"].(float64); ok {
+			geoLocation["lat"] = existingLat
+		}
+		if existingLon, ok := p.json["longitude"].(float64); ok {
+			geoLocation["lon"] = existingLon
 		}
 	}
+
+	p.json["geolocation"] = geoLocation
+
 	return nil
-}
-
-// repackageGeolocation reformats the geolocation field as a map with keys "lat" and "lon".
-func (p *Profile) repackageGeolocation() {
-	if p.json["latitude"] != nil || p.json["longitude"] != nil {
-		geoLocation := make(map[string]interface{})
-
-		if p.json["latitude"] != nil {
-			geoLocation["lat"] = p.json["latitude"]
-		} else {
-			geoLocation["lat"] = 0
-		}
-
-		if p.json["longitude"] != nil {
-			geoLocation["lon"] = p.json["longitude"]
-		} else {
-			geoLocation["lon"] = 0
-		}
-
-		p.json["geolocation"] = geoLocation
-	}
 }
 
 // normalizeCountryCode takes in profile JSON and normalizes the country information present.

--- a/services/index/internal/model/profile_test.go
+++ b/services/index/internal/model/profile_test.go
@@ -71,22 +71,80 @@ func TestConvertGeolocation(t *testing.T) {
 				"geolocation": "40.748817,-73.985428"
 			}`,
 			expected: map[string]interface{}{
-				"geolocation": "40.748817,-73.985428",
-				"latitude":    40.748817,
-				"longitude":   -73.985428,
+				"geolocation": map[string]interface{}{
+					"lat": 40.748817,
+					"lon": -73.985428,
+				},
 			},
 			expectErr: false,
 		},
 		{
-			name:      "invalid geolocation string",
-			input:     `{"geolocation": "invalid_string"}`,
-			expected:  map[string]interface{}{"geolocation": "invalid_string"},
+			name: "only latitude present",
+			input: `{
+				"latitude": 40.748817
+			}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{
+					"lat": 40.748817,
+					"lon": 0,
+				},
+			},
+			expectErr: false,
+		},
+		{
+			name: "only longitude present",
+			input: `{
+				"longitude": -73.985428
+			}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{
+					"lat": 0,
+					"lon": -73.985428,
+				},
+			},
+			expectErr: false,
+		},
+		{
+			name: "both latitude and longitude present",
+			input: `{
+				"latitude": 40.748817,
+				"longitude": -73.985428
+			}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{
+					"lat": 40.748817,
+					"lon": -73.985428,
+				},
+			},
+			expectErr: false,
+		},
+		{
+			name: "geolocation as object",
+			input: `{
+				"geolocation": {"lat": 40.748817, "lon": -73.985428}
+			}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{
+					"lat": 40.748817,
+					"lon": -73.985428,
+				},
+			},
+			expectErr: false,
+		},
+		{
+			name:  "invalid geolocation string",
+			input: `{"geolocation": "invalid_string"}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{"lat": 0, "lon": 0},
+			},
 			expectErr: true,
 		},
 		{
-			name:      "geolocation not present",
-			input:     `{}`,
-			expected:  map[string]interface{}{},
+			name:  "geolocation not present",
+			input: `{}`,
+			expected: map[string]interface{}{
+				"geolocation": map[string]interface{}{"lat": 0, "lon": 0},
+			},
 			expectErr: false,
 		},
 	}
@@ -101,69 +159,6 @@ func TestConvertGeolocation(t *testing.T) {
 				require.NoError(t, err)
 				require.Equal(t, tt.expected, profile.GetJSON())
 			}
-		})
-	}
-}
-
-func TestRepackageGeolocation(t *testing.T) {
-	tests := []struct {
-		name     string
-		input    string
-		expected map[string]interface{}
-	}{
-		{
-			name: "both latitude and longitude are present",
-			input: `{
-				"latitude":  40.748817,
-				"longitude": -73.985428
-			}`,
-			expected: map[string]interface{}{
-				"latitude":  40.748817,
-				"longitude": -73.985428,
-				"geolocation": map[string]interface{}{
-					"lat": 40.748817,
-					"lon": -73.985428,
-				},
-			},
-		},
-		{
-			name: "only latitude is present",
-			input: `{
-				"latitude": 40.748817
-			}`,
-			expected: map[string]interface{}{
-				"latitude": 40.748817,
-				"geolocation": map[string]interface{}{
-					"lat": 40.748817,
-					"lon": 0,
-				},
-			},
-		},
-		{
-			name: "only longitude is present",
-			input: `{
-				"longitude": -73.985428
-			}`,
-			expected: map[string]interface{}{
-				"longitude": -73.985428,
-				"geolocation": map[string]interface{}{
-					"lat": 0,
-					"lon": -73.985428,
-				},
-			},
-		},
-		{
-			name:     "neither latitude nor longitude are present",
-			input:    `{}`,
-			expected: map[string]interface{}{},
-		},
-	}
-
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			profile := model.NewTestProfile(tt.input)
-			profile.RepackageGeolocation()
-			require.Equal(t, tt.expected, profile.GetJSON())
 		})
 	}
 }

--- a/services/index/internal/model/profile_test.go
+++ b/services/index/internal/model/profile_test.go
@@ -10,6 +10,54 @@ import (
 	"github.com/MurmurationsNetwork/MurmurationsServices/services/index/internal/model"
 )
 
+func TestGetJSON(t *testing.T) {
+	tests := []struct {
+		name     string
+		profile  *model.Profile
+		expected map[string]interface{}
+	}{
+		{
+			name: "Test with allowed fields",
+			profile: model.NewProfile(`{
+				"name": "John",
+				"geolocation": "40.7128,-74.0060",
+				"last_updated": 1630843200,
+				"linked_schemas": "schema1",
+				"country": "USA"
+			}`),
+			expected: map[string]interface{}{
+				"name":           "John",
+				"geolocation":    "40.7128,-74.0060",
+				"last_updated":   float64(1630843200),
+				"linked_schemas": "schema1",
+				"country":        "USA",
+			},
+		},
+		{
+			name: "Test with disallowed fields",
+			profile: model.NewProfile(`{
+				"name": "John",
+				"garbageField": "garbageValue"
+			}`),
+			expected: map[string]interface{}{
+				"name": "John",
+			},
+		},
+		{
+			name:     "Test with empty profile",
+			profile:  model.NewProfile(`{}`),
+			expected: map[string]interface{}{},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			actual := tt.profile.GetJSON()
+			require.Equal(t, tt.expected, actual)
+		})
+	}
+}
+
 func TestConvertGeolocation(t *testing.T) {
 	tests := []struct {
 		name      string

